src/test/java/org/jenkinsci/plugins/workflow/flow/FlowExecutionListTest.java

```diff
@@ -160,6 +164,34 @@ public class FlowExecutionListTest {
         });
     }
 
+    @Test public void stepExecutionIteratorDoesNotLeakBuildsWhenOneIsStuck() throws Throwable {
+        sessions.then(r -> {
+            var notStuck = r.createProject(WorkflowJob.class, "not-stuck");
+            notStuck.setDefinition(new CpsFlowDefinition("semaphore 'wait'", true));
+            var notStuckBuild = notStuck.scheduleBuild2(0).waitForStart();
+            SemaphoreStep.waitForStart("wait/1", notStuckBuild);
+            WeakReference<Object> notStuckBuildRef = new WeakReference<>(notStuckBuild);
+            // Create a Pipeline that runs a long-lived task on its CpsVmExecutorService, causing it to get stuck.
+            var stuck = r.createProject(WorkflowJob.class, "stuck");
+            stuck.setDefinition(new CpsFlowDefinition("blockSynchronously 'stuck'", false));
+            var stuckBuild = stuck.scheduleBuild2(0).waitForStart();
+            await().atMost(5, TimeUnit.SECONDS).until(() -> SynchronousBlockingStep.isStarted("stuck"));
+            // Make FlowExecutionList$StepExecutionIteratorImpl.applyAll submit a task to the CpsVmExecutorService
+            // for stuck #1 that will never complete, so the resulting future will never complete.
+            StepExecution.applyAll(e -> null);
```

Comment:
Should this return value be kept in a local variable?

Reply:
It's not necessary for the leak, if that's what you mean. Even if the `StepExecution.applyAll` caller completely ignores the return value, the stuck CPS VM thread holds a strong reference to the aggregate future because the `SingleLaneExecutorService`'s task queue has a reference to the `SettableFuture` for the `getCurrentExecutions` call, and then that future references the aggregate future via a listener just due to the implementation of `Futures.allAsList`.